File: leetcode/DinnerPlates.cpp
/**********************************************
  > File Name		: DinnerPlates.cpp
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Sun 03 Apr 2022 02:47:46 PM CST
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

#include <vector>
#include <iostream>
#include <stack>
#include <set>
using namespace std;

class DinnerPlates {
private:
    vector<stack<int>> plates;
    set<int> remain, nonempty;
    int capacity;
public:
    DinnerPlates(int capacity): capacity(capacity) {}
    
    void push(int val) {
        if (this->remain.empty()) {
            this->plates.push_back(stack<int>({val}));
            if (this->capacity > 1)
                this->remain.insert(this->plates.size()-1);
            this->nonempty.insert(this->plates.size()-1);
        } else {
            auto minRemain = this->remain.begin();
            stack<int>& s = this->plates[*minRemain];
            s.push(val);

            if (s.size() == 1) {
                this->nonempty.insert(*minRemain);
            }

            if (s.size() == this->capacity) {
                this->remain.erase(minRemain);
            }
        }
    }
    
    int pop() {
        if (this->nonempty.empty())
            return -1;
        auto maxNonempty = --this->nonempty.end();
        stack<int>& s = this->plates[*maxNonempty];

        int res = s.top(); s.pop();

        if (s.size() == this->capacity-1) {
            this->remain.insert(*maxNonempty);
        }

        if (s.empty()) {
            this->nonempty.erase(maxNonempty);
        }
        return res;
    }
    
    int popAtStack(int index) {
        if (index >= this->plates.size() || this->plates[index].empty())
            return -1;

        stack<int>& s = this->plates[index];
        int res = s.top(); s.pop();

        if (s.size() == this->capacity-1)
            this->remain.insert(index);

        if (s.empty()) {
            this->nonempty.erase(this->nonempty.find(index));
        }
        return res;
    }
};

int main() {
    set<int> s = {4, 6, 1, 2};
    cout << *s.begin() << endl;
    cout << *s.rbegin() << endl;
}

/**
 * Your DinnerPlates object will be instantiated and called as such:
 * DinnerPlates* obj = new DinnerPlates(capacity);
 * obj->push(val);
 * int param_2 = obj->pop();
 * int param_3 = obj->popAtStack(index);
 */

File: leetcode/DinnerPlates.rs
/**********************************************
  > File Name		: DinnerPlates.rs
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Sun 03 Apr 2022 11:18:30 PM CST
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

use std::collections::BTreeSet;

struct DinnerPlates {
    capacity: usize,
    plates: Vec<Vec<i32>>,
    remain: BTreeSet<usize>,
    nonempty: BTreeSet<usize>
}


/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl DinnerPlates {

    fn new(capacity: i32) -> Self {
        Self {
            capacity: capacity as usize,
            plates: Vec::new(),
            remain: BTreeSet::new(),
            nonempty: BTreeSet::new()
        }
    }
    
    fn push(&mut self, val: i32) {
        if self.remain.is_empty() {
            self.plates.push(vec![val]);
            let size = self.plates.len();
            
            self.nonempty.insert(size-1);
            if self.capacity > 1 {
                self.remain.insert(size-1);
            }
        } else {
            let min_remain: usize = *self.remain.iter().min().unwrap();
            let stack_size = {
                let s = &mut self.plates[min_remain];
                s.push(val);
                s.len()
            };
            
            if stack_size == self.capacity {
                self.remain.remove(&min_remain);
            }

            if stack_size == 1 {
                self.nonempty.insert(min_remain);
            }
        }
    }
    
    fn pop(&mut self) -> i32 {
        if self.nonempty.is_empty() {
            return -1;
        }

        let max_nonempty: usize = *self.nonempty.iter().max().unwrap();
        let (res, stack_size) = {
            let s = &mut self.plates[max_nonempty];
            (s.pop().unwrap(), s.len())
        };

        if stack_size == self.capacity-1 {
            self.remain.insert(max_nonempty);
        }

        if stack_size == 0 {
            self.nonempty.remove(&max_nonempty);
        }
        res
    }
    
    fn pop_at_stack(&mut self, i: i32) -> i32 {
        let index = i as usize;
        if self.nonempty.is_empty() || index >= self.plates.len() {
            return -1;
        }

        let (res, stack_size) = {
            let s = &mut self.plates[index];
            if s.is_empty() {
                return -1;
            }
            (s.pop().unwrap(), s.len())
        };

        if stack_size == self.capacity-1 {
            self.remain.insert(index);
        }

        if stack_size == 0 {
            self.nonempty.remove(&index);
        }

        res
    }
}

fn main() {
    let mut a = BTreeSet::new();
    a.insert(5); a.insert(2); a.insert(7); a.insert(1);
    println!("{:?}", a.iter().max().unwrap());
}

/*
 * Your DinnerPlates object will be instantiated and called as such:
 * let obj = DinnerPlates::new(capacity);
 * obj.push(val);
 * let ret_2: i32 = obj.pop();
 * let ret_3: i32 = obj.pop_at_stack(index);
 */
